### BUSINESS UNDERSTANDING 
## Overview
 The Lowest seasons of  human kind  are those of global health  pandemics with the  most recent being Covid _19.Preceeded by the other influenza viruses the H1 _N1 and the seasonal flu.According to the World Health Organization , There are around a billion cases of seasonal influenza annually, including 3–5 million cases of severe illness.
It causes 290 000 to 650 000 respiratory deaths annually.
Ninety-nine percent of deaths in children under 5 years of age with influenza-related lower respiratory tract infections are in developing countries https://www.who.int/news-room/fact-sheets/detail/influenza-(seasonal)?gad_source=1&gclid=EAIaIQobChMIm-y9mLnAhgMV5KmDBx1GgwYuEAA


The flu and H1N1 vaccination rate in the United States is relatively low at 50.2%. This presents a public 
health concern, particularly given the high availability and proven effectiveness of flu vaccines.As the world struggles to vaccinate the global population against COVID-19, an understanding of how people’s backgrounds, opinions, and health behaviors are related to their personal vaccination patterns can provide guidance for future public health efforts.

Understanding these factors and modelling them  as a data scientist will be pertinent in infroming  the the public health stakeholders in the  the World Health Organization (WHO).

## Proposed Solution
To develop a machine learning model that can predict the likelihood of an individual getting vaccinated by perfoming an analysis of the data provided by  drivendata.org.Performance will be evaluated according to the area under the receiver operating characteristic curve (ROC AUC) for each of the two target variables. The mean of these two scores will be the overall score. A higher value indicates stronger performance.

## Challenges 
 One of the setbacks of this project would be biasness ,because the data used was done through a survey of the USA.Therefore the model may not fit in the developing countries because the factors that are taken into account are different. 


 ### Problem Statement
 The goal is to determine the most effective behaviors that lead people to get vaccinated against the H1N1 flu virus and seasonal flu viruses. This involves identifying key factors that influence behavior and motivations, such as education, age, employer influence, health insurance, physician recommendations, presence of chronic diseases, and employment in the healthcare sector. Additionally, vaccination status may be affected by behaviors like face-touching, attending large gatherings, wearing masks, and hand-washing habits.

The primary objective of this algorithm is classification. It should analyze attributes from the National 2009 H1N1 Flu and Seasonal Flu Survey dataset to determine if a respondent received the seasonal flu vaccine or the H1N1 flu vaccine. The model's effectiveness will be measured by how accurately it predicts vaccination status based on the respondent's behaviors and motivational factors.


## Objectives 
. To identify key factors that influence people to be vaccinated.


.To identify  the age group that is majorly affected by the vaccination


. To develop a machine learning algorithim that predicts vaccination status based on respondent's behaviours.




## Data Understanding 
Data Sources

The data comes from the National 2009 H1N1 Flu Survey (NHFS) and contains answers from respondents regarding demographic information, opinions on vaccinations, and behavioral information.

The data was collected via a list-assisted random-digit-dialing telephone survey of households. The target population was all persons 6 months or older living in the United States. The survey produced estimates of vaccine coverage rates for both the monovalent pH1N1 and trivalent seasonal influenza vaccines.
 
DrivenData. (n.d.). Flu Shot Learning: Predict H1N1 and Seasonal Flu Vaccines. DrivenData from https://www.drivendata.org/competitions/66/flu-shot-learning/data/





## variable Explanation

For this dataset, there are two target variables:

h1n1_vaccine - Whether respondent received H1N1 flu vaccine.

seasonal_vaccine - Whether respondent received seasonal flu vaccine.

Both are binary variables: 0 = No; 1 = Yes. Some respondents didn't get either vaccine, others got only one, and some got both. This is formulated as a multilabel (and not multiclass) problem.




## Column Description
The first column respondent_id is a unique and random identifier. For all binary variables: 0 = No; 1 = Yes.

h1n1_concern - Level of concern about the H1N1 flu. 0 = Not at all concerned; 1 = Not very concerned; 2 = Somewhat concerned; 3 = Very concerned.

h1n1_knowledge - Level of knowledge about H1N1 flu. 0 = No knowledge; 1 = A little knowledge; 2 = A lot of knowledge.

behavioral_antiviral_meds - Has taken antiviral medications. (binary)

behavioral_avoidance - Has avoided close contact with others with flu-like symptoms. (binary) behavioral_face_mask - Has bought a face mask. (binary)

behavioral_wash_hands - Has frequently washed hands or used hand sanitizer. (binary)

behavioral_large_gatherings - Has reduced time at large gatherings. (binary)

behavioral_outside_home - Has reduced contact with people outside of own household. (binary)

behavioral_touch_face - Has avoided touching eyes, nose, or mouth. (binary)

doctor_recc_h1n1 - H1N1 flu vaccine was recommended by doctor. (binary)

doctor_recc_seasonal - Seasonal flu vaccine was recommended by doctor. (binary)

chronic_med_condition - Has any of the following chronic medical conditions: asthma or an other lung condition, diabetes, a heart condition, a kidney condition, sickle cell anemia or other anemia, a neurological or neuromuscular condition, a liver condition, or a weakened immune system caused by a chronic illness or by medicines taken for a chronic illness. (binary)

child_under_6_months- Has regular close contact with a child under the age of six months. (binary)

health_worker- Is a healthcare worker. (binary)

health_insurance- Has health insurance. (binary)

opinion_h1n1_vacc_effective - Respondent's opinion about H1N1 vaccine effectiveness. 1 = Not at all effective; 2 = Not very effective; 3 = Don't know; 4 = Somewhat effective; 5 = Very effective.

opinion_h1n1_risk - Respondent's opinion about risk of getting sick with H1N1 flu without vaccine. 1 = Very Low; 2 = Somewhat low; 3 = Don't know; 4 = Somewhat high; 5 = Very high.

opinion_h1n1_sick_from_vacc - Respondent's worry of getting sick from taking H1N1 vaccine. 1 = Not at all worried; 2 = Not very worried; 3 = Don't know; 4 = Somewhat worried; 5 = Very worried.

opinion_seas_vacc_effective - Respondent's opinion about seasonal flu vaccine effectiveness. 1 = Not at all effective; 2 = Not very effective; 3 = Don't know; 4 = Somewhat effective; 5 = Very effective.

opinion_seas_risk - Respondent's opinion about risk of getting sick with seasonal flu without vaccine. 1 = Very Low; 2 = Somewhat low; 3 = Don't know; 4 = Somewhat high; 5 = Very high.

opinion_seas_sick_from_vacc - Respondent's worry of getting sick from taking seasonal flu vaccine. 1 = Not at all worried; 2 = Not very worried; 3 = Don't know; 4 = Somewhat worried; 5 = Very worried.

age_group - Age group of respondent.

education - Self-reported education level.

race - Race of respondent.

sex- Sex of respondent.

income_poverty - Household annual income of respondent with respect to 2008 Census poverty thresholds.

marital_status - Marital status of respondent.

rent_or_own - Housing situation of respondent.

employment_status- Employment status of respondent.

hhs_geo_region - Respondent's residence using a 10-region geographic classification defined by the U.S. Dept. of Health and Human Services. Values are represented as short random character strings.

census_msa - Respondent's residence within metropolitan statistical areas (MSA) as defined by the U.S. Census.

household_adults - Number of other adults in household, top-coded to 3.

household_children - Number of children in household, top-coded to 3.

employment_industry - Type of industry respondent is employed in. Values are represented as short random character strings.

employment_occupation - Type of occupation of respondent. Values are represented as short random character strings.


### Data Preparation & Exploration

In [6]:
#Import relevant libraries
# Import Python libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

The code below is a function that loads and merges the feature csv file and the labels csv file.

In [41]:
#define funtion to load,describe &give shape of data

def load_and_merge_csv(file_path1, file_path2, merge_key):
    # Load the CSV files
    df1 = pd.read_csv(file_path1)
    df2 = pd.read_csv(file_path2)
    
    # Merge the dataframes on the specified key
    merged_df = pd.merge(df1, df2, on=merge_key)
    
    # Get the shape of the merged dataframe
    shape = merged_df.shape
    
    # Get information about the merged dataframe
    info = merged_df.info()
    
    # Get descriptive statistics of the merged dataframe
    description = merged_df.describe()
    
    return merged_df, shape, info, description


file_path1 = 'H1 flu/training_set_features.csv'
file_path2 = 'H1 flu/training_set_labels.csv'
merge_key = 'respondent_id'  # column name  to merge on
merged_df, shape, info, description = load_and_merge_csv(file_path1, file_path2, merge_key)

print("Shape of the merged dataframe:", shape)
print("\nInfo of the merged dataframe:")
print(info)
print("\nDescription of the merged dataframe:")
print(description)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26707 entries, 0 to 26706
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26707 non-null  int64  
 1   h1n1_concern                 26615 non-null  float64
 2   h1n1_knowledge               26591 non-null  float64
 3   behavioral_antiviral_meds    26636 non-null  float64
 4   behavioral_avoidance         26499 non-null  float64
 5   behavioral_face_mask         26688 non-null  float64
 6   behavioral_wash_hands        26665 non-null  float64
 7   behavioral_large_gatherings  26620 non-null  float64
 8   behavioral_outside_home      26625 non-null  float64
 9   behavioral_touch_face        26579 non-null  float64
 10  doctor_recc_h1n1             24547 non-null  float64
 11  doctor_recc_seasonal         24547 non-null  float64
 12  chronic_med_condition        25736 non-null  float64
 13  child_under_6_mo

In [38]:
# View the dataframe 
merged_df.head(10)


h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0           1.0             0.0                        0.0   
1           3.0             2.0                        0.0   
2           1.0             1.0                        0.0   
3           1.0             1.0                        0.0   
4           2.0             1.0                        0.0   
5           3.0             1.0                        0.0   
6           0.0             0.0                        0.0   
7           1.0             0.0                        0.0   
8           0.0             2.0                        0.0   
9           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   
5                   1.0                   0.0                    1.0   
6                   0.0                   0.0                    0.0   
7                   1.0                   0.0                    1.0   
8                   1.0                   0.0                    1.0   
9                   1.0                   0.0                    0.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   
5                          0.0                      0.0   
6                          0.0                      0.0   
7                          0.0                      0.0   
8                          1.0                      1.0   
9                          1.0                      0.0   

   behavioral_touch_face  doctor_recc_h1n1  ...  rent_or_own  \
0                    1.0               0.0  ...          Own   
1                    1.0               0.0  ...         Rent   
2                    0.0               NaN  ...          Own   
3                    0.0               0.0  ...         Rent   
4                    1.0               0.0  ...          Own   
5                    1.0               0.0  ...          Own   
6                    0.0               0.0  ...          Own   
7                    1.0               1.0  ...          Own   
8                    1.0               0.0  ...          Own   
9                    1.0               0.0  ...          Own   

    employment_status  hhs_geo_region                census_msa  \
0  Not in Labor Force        oxchjgsf                   Non-MSA   
1            Employed        bhuqouqj  MSA, Not Principle  City   
2            Employed        qufhixun  MSA, Not Principle  City   
3  Not in Labor Force        lrircsnp       MSA, Principle City   
4            Employed        qufhixun  MSA, Not Principle  City   
5            Employed        atmpeygn       MSA, Principle City   
6            Employed        qufhixun  MSA, Not Principle  City   
7            Employed        bhuqouqj                   Non-MSA   
8            Employed        bhuqouqj  MSA, Not Principle  City   
9  Not in Labor Force        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   
5               2.0                 3.0           

Data Quality check


   .Identify missing values,null values and duplicates in the merged dataset

In [42]:
def check_data_quality(merged_df):
    # Check for missing values
    missing_values = merged_df.isnull().sum()
    print("Missing values in each column:\n", missing_values)
    
    # Check for duplicate rows
    duplicate_rows =merged_df.duplicated().sum()
    print("Number of duplicate rows:", duplicate_rows)
    
    return missing_values, duplicate_rows
# Check data quality
print("Missing values in each column:\n", missing_values)
print("Number of duplicate rows:", duplicate_rows)

Missing values in each column:
 h1n1_concern                      92
h1n1_knowledge                   116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_h1n1                2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_h1n1_vacc_effective      391
opinion_h1n1_risk                388
opinion_h1n1_sick_from_vacc      395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4423
marita

The output above shows :

Columns like health_insurance have a very high number of missing values (12,274), indicating significant data loss in that column.


Columns doctor_recc_h1n1 and doctor_recc_seasonal also have a large number of missing values (2,160 each).


Several columns, such as age_group, race, h1n1_vaccine, and seasonal_vaccine, have no missing values, indicating complete data for these columns.

There are no duplicate rows in the DataFrame, indicating that each record in the dataset is unique.


For Analysis the target variable will be the h1n1 vaccine column and the other target variable will be seasonal  vaccine 

In [44]:
#Assign H1n1 as target variable and season flu as other target variable

target_variable = 'h1n1_vaccine'

other_target = 'seasonal_vaccine'



In [14]:
# Dealing with missing values
 